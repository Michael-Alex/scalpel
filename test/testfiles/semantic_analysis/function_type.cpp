typedef int fun_t();

fun_t f;
fun_t* f_ptr;

class c
{
	typedef int const_fun_t() const;

	fun_t f;
	static fun_t g;
	const_fun_t h;
};

int
c::f()
{
	return 0;
}

