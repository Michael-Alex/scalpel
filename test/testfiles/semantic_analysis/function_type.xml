<namespace id="n0">
 <typedef name="predicate">
  <type>
   <function_type>
    <return_type>
     <fundamental_type type="bool"/>
    </return_type>
    <parameter_types>
     <fundamental_type type="int"/>
     <fundamental_type type="int"/>
    </parameter_types>
   </function_type>
  </type>
 </typedef>
</namespace>
