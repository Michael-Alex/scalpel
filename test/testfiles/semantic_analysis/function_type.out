{
  "id": "namespace-0",
  "classes":
  [
    {
      "id": "class-0",
      "name": "c",
      "typedefs":
      [
        {
          "id": "member-typedef-0",
          "name": "const_fun_t",
          "access": "private",
          "type":
          {
            "function type":
            {
              "const": true,
              "return type":
              {
                "fundamental type": "int"
              },
              "parameter types":
              [
              ]
            }
          }
        }
      ],
      "destructor":
      {
        "access": "public"
      },
      "simple functions":
      [
        {
          "id": "simple-member-function-0",
          "name": "f",
          "access": "private",
          "return type":
          {
            "fundamental type": "int"
          }
        },
        {
          "id": "simple-member-function-1",
          "name": "g",
          "access": "private",
          "defined": false,
          "return type":
          {
            "fundamental type": "int"
          }
        },
        {
          "id": "simple-member-function-2",
          "name": "h",
          "access": "private",
          "const": true,
          "defined": false,
          "return type":
          {
            "fundamental type": "int"
          }
        }
      ]
    }
  ],
  "typedefs":
  [
    {
      "id": "typedef-0",
      "name": "fun_t",
      "type":
      {
        "function type":
        {
          "return type":
          {
            "fundamental type": "int"
          },
          "parameter types":
          [
          ]
        }
      }
    }
  ],
  "simple functions":
  [
    {
      "id": "simple-function-0",
      "name": "f",
      "defined": false,
      "return type":
      {
        "fundamental type": "int"
      }
    },
    {
      "id": "simple-function-1",
      "name": "g",
      "defined": false,
      "return type":
      {
        "fundamental type": "void"
      },
      "parameters":
      [
        {
          "type":
          {
            "function type":
            {
              "return type":
              {
                "fundamental type": "int"
              },
              "parameter types":
              [
              ]
            }
          }
        }
      ]
    },
    {
      "id": "simple-function-2",
      "name": "h",
      "defined": false,
      "return type":
      {
        "fundamental type": "void"
      },
      "parameters":
      [
        {
          "type":
          {
            "function type":
            {
              "variadic": true,
              "return type":
              {
                "pointer":
                {
                  "type":
                  {
                    "fundamental type": "int"
                  }
                }
              },
              "parameter types":
              [
                {
                  "fundamental type": "unsigned int"
                }
              ]
            }
          }
        }
      ]
    }
  ],
  "variables":
  [
    {
      "id": "variable-0",
      "name": "f_ptr",
      "type":
      {
        "pointer":
        {
          "type":
          {
            "function type":
            {
              "return type":
              {
                "fundamental type": "int"
              },
              "parameter types":
              [
              ]
            }
          }
        }
      }
    }
  ]
}
